shared/src/main/scala/mlscript/TypeDefs.scala

```diff
@@ -555,89 +555,92 @@ class TypeDefs extends ConstraintSolver { self: Typer =>
     var varianceUpdated: Bool = false;
 
     def fieldVarianceHelper(fieldTy: FieldType, curVariance: VarianceInfo)(implicit ctx: Ctx, tyDef: TypeDef): Unit = {
-        updateVariance(fieldTy.ub, curVariance.flip(VarianceInfo.co()))
-        fieldTy.lb.foreach(lb => updateVariance(lb, curVariance.flip(VarianceInfo.contra())))
+        fieldTy.lb.foreach(lb => updateVariance(lb, curVariance.flip))
+        updateVariance(fieldTy.ub, curVariance)
     }
 
-    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit ctx: Ctx, tyDef: TypeDef): Unit = ty match {
-      case ProxyType(underlying) => updateVariance(underlying, curVariance)
-      case TraitTag(_) | ClassTag(_, _) => ()
-      case ExtrType(pol) => ()
-      case t: TypeVariable => tyDef.tvarVariance.updateWith(t) {
-        case None => Some(curVariance)
-        case Some(value) =>
-          val newVariance = value && curVariance
-          if (newVariance =/= curVariance) varianceUpdated = true;
-          Some(newVariance)
-      })
-      case RecordType(fields) => fields.foreach {
-        case (_ , fieldTy) => fieldVarianceHelper(fieldTy, curVariance)
-      }
-      case NegType(negated) =>
-        updateVariance(negated, curVariance.flip(VarianceInfo.contra()))
-      case TypeRef(defn, targs) =>
-        val typeRefDef = ctx.tyDefs.getOrElse(defn.name, throw new Exception(s"Cannot find $defn type definition in context"))
-        targs.zip(typeRefDef.tvars).foreach { case (targ, tvar) =>
-          val tvarVariance = if (curVariance === VarianceInfo.contra())
-              typeRefDef.tvarVariance.getOrElse(tvar, VarianceInfo.bi()).flip(VarianceInfo.contra())
-            else
-              typeRefDef.tvarVariance.getOrElse(tvar, VarianceInfo.bi()).flip(VarianceInfo.co())
-          updateVariance(targ, tvarVariance)
-        }
-      case ComposedType(pol, lhs, rhs) =>
-        updateVariance(lhs, curVariance)
-        updateVariance(rhs, curVariance)
-      case TypeBounds(lb, ub) =>
-        updateVariance(lb, curVariance.flip(VarianceInfo.contra()))
-        updateVariance(ub, curVariance.flip(VarianceInfo.co()))
-      case ArrayType(inner) => fieldVarianceHelper(inner, curVariance)
-      case TupleType(fields) => fields.foreach {
-          case (_ , fieldTy) => fieldVarianceHelper(fieldTy, curVariance)
+    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit ctx: Ctx, tyDef: TypeDef): Unit =
```

Comment:
No need to pass `ctx`. It can already be captured from the outer scope.
```suggestion
    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit tyDef: TypeDef): Unit =
```